examples/with-metadata.rs

```diff
@@ -0,0 +1,146 @@
+//! A single threaded executor that uses shortest-job-first scheduling.
+
+use std::cell::RefCell;
+use std::collections::BinaryHeap;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::thread;
+use std::time::{Duration, Instant};
+use std::{cell::Cell, future::Future};
+
+use async_task::{Builder, Runnable, Task};
+use pin_project_lite::pin_project;
+use smol::{channel, future};
+
+struct ByDuration(Runnable<DurationMetadata>);
+
+impl PartialEq for ByDuration {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.metadata().inner.get() == other.0.metadata().inner.get()
+    }
+}
+
+impl Eq for ByDuration {}
+
+impl PartialOrd for ByDuration {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.0
+            .metadata()
+            .inner
+            .get()
+            .partial_cmp(&other.0.metadata().inner.get())
+            .map(|ord| ord.reverse())
+    }
+}
+
+impl Ord for ByDuration {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.0
+            .metadata()
+            .inner
+            .get()
+            .cmp(&other.0.metadata().inner.get())
+            .reverse()
+    }
+}
+
+pin_project! {
+    #[must_use = "futures do nothing unless you `.await` or poll them"]
+    struct MeasureRuntime<'a, F> {
+        #[pin]
+        f: F,
+        duration: &'a Cell<Duration>
+    }
+}
+
+impl<'a, F: Future> Future for MeasureRuntime<'a, F> {
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        let duration_cell = this.duration;
+        let start = Instant::now();
+        let res = F::poll(this.f, cx);
+        let new_duration = Instant::now() - start;
+        duration_cell.set(duration_cell.get() / 2 + new_duration / 2);
+        res
+    }
+}
```

Comment:
This could be expressed using `smol::future::poll_fn` instead of  manually.

Reply:
I'm trying to imagine how it would work, but it seems awkward compared to the manual impl.

Reply:
Like this, I'd imagine:

```rust
let my_func = move |metadata| {
    let my_future = smol::pin!(my_future);
    future::poll_fn(|cx| {
        let start = Instant::now();
        let res = my_future.poll(cx);
        let new_duration = Instant::now() - start;
        metadata.duration.set(metadata.duration.get() / 2 + new_duration / 2);
        res
    });
};
```

If the borrow checker doesn't like it, feel free to stick with what you have.

Reply:
I'm fairly sure the borrow checker will be very angry with that. The `Pin<&mut Fut>` has to be moved into the poll function, but then the actual `Fut` value has to be moved too. The only alternative I can see is doing `pin!` inside the poll function instead, and maybe that works?

Reply:
Yep, borrow checker is very upset with trying to use `poll_fn` here.